File: apps/ios/TelarMobile/Stores/UsageReport.swift
import Foundation

/// SPEND OVER TIME — the wire shape `GET /api/usage` answers with, and the fold
/// the screen renders. Mirrors `UsageReport` in
/// packages/engine-client/src/protocol/common.ts and `foldUsage` in
/// apps/web/lib/usage-report.ts.
///
/// THE REPORT IS ONE MAC'S. It is derived by scanning that machine's provider
/// transcripts (`~/.claude/projects`, `~/.codex/sessions`), so there is no such
/// thing as a merged figure across paired Macs: two of them are two ledgers of
/// two machines' work, and adding them would answer a question nobody asked.
/// The screen picks a Mac; it never sums.
///
/// COST IS THE PROVIDER'S FIGURE where the transcript carries one and the rate
/// table's otherwise. A model neither knows stays UNPRICED: its tokens count and
/// its cost reads as absent, never as $0.00 — a zero would claim the work was
/// free.
extension TokenUsage {
    static let zero = TokenUsage(input: 0, output: 0, cacheRead: 0, cacheCreate: 0)

    /// Every token the window moved — the screen's headline figure, and the
    /// web's `processedTokens`. `reasoning` is deliberately absent from the sum:
    /// the providers that report it also count it inside `output`, so adding it
    /// would count those tokens twice.
    var processed: Int { input + output + cacheRead + cacheCreate }
}

struct UsageBucket: Decodable, Equatable {
    /// `YYYY-MM-DD` in the requested zone for days; the hour-start epoch ms as a
    /// decimal string for hours.
    var period: String
    var driver: String
    var model: String
    var tokens: TokenUsage
    var costUsd: Double
    /// Whether every record here has a cost — provider-reported or rate-priced.
    var priced: Bool
    /// Records, not turns: one Claude assistant message or one Codex token
    /// count. The screen says "requests" for this reason.
    var turns: Int
}

/// One transcript directory's scan outcome, so the screen can say what was and
/// was not counted rather than letting a missing install read as zero use.
struct UsageSource: Decodable, Equatable {
    var provider: String
    /// ok | missing | failed.
    var status: String
    var path: String
    var files: Int
    var sessions: Int
}

struct UsageReport: Decodable, Equatable {
    var sinceMs: Timestamp
    var untilMs: Timestamp
    /// day | hour.
    var resolution: String
    var timeZone: String
    var buckets: [UsageBucket]
    var sources: [UsageSource]
    /// fresh | cached | unavailable — where rate-priced costs came from.
    var pricing: String
    var sessions: Int
    var readAt: Timestamp

    static let empty = UsageReport(sinceMs: 0, untilMs: 0, resolution: "day", timeZone: "UTC",
                                   buckets: [], sources: [], pricing: "unavailable", sessions: 0, readAt: 0)
}

// MARK: - the fold

/// One slice's totals — the web's `UsageTotals`.
struct UsageTotals: Equatable {
    var tokens = TokenUsage.zero
    var processed = 0
    var costUsd: Double = 0
    /// True when every counted record carried a cost figure.
    var priced = true
    var turns = 0

    mutating func fold(_ bucket: UsageBucket) {
        tokens.input += bucket.tokens.input
        tokens.output += bucket.tokens.output
        tokens.cacheRead += bucket.tokens.cacheRead
        tokens.cacheCreate += bucket.tokens.cacheCreate
        processed += bucket.tokens.processed
        costUsd += bucket.costUsd
        priced = priced && bucket.priced
        turns += bucket.turns
    }
}

struct UsageProviderSlice: Identifiable, Equatable {
    var driver: String
    var totals: UsageTotals
    /// This provider's share of PROCESSED TOKENS — never of cost. Cost is absent
    /// for a whole provider (Codex reports none), and a share of a figure
    /// missing half its terms would always read 100% Claude. The web's rule.
    var share: Double
    var id: String { driver }
}

struct UsageFold: Equatable {
    var total = UsageTotals()
    var providers: [UsageProviderSlice] = []
    var sessions = 0
}

/// THE DRIVERS, IN THE ORDER THE WEB LISTS THEM, and their labels — the phone
/// and the Mac name the same provider the same way.
let usageDrivers = ["claude", "codex", "opencode"]

func usageDriverLabel(_ driver: String) -> String {
    switch driver {
    case "claude": "Claude"
    case "codex": "Codex"
    case "opencode": "OpenCode"
    // A provider this build predates is named by its own id rather than
    // dropped: an unlabelled row still carries real spend.
    default: driver
    }
}

/// The screen's one derivation — `foldUsage`, minus the per-period and per-model
/// slices the phone's screen does not draw (there is no chart and no breakdown
/// table here, by #404's own scope). Pure, and tested.
///
/// A PROVIDER WITH NOTHING IN THE WINDOW IS NOT A ROW. The web's rule: a Codex
/// line reading 0% / $0.00 on a week nobody ran Codex is a fact about the
/// provider list, not about the window.
func foldUsage(_ report: UsageReport) -> UsageFold {
    var total = UsageTotals()
    var byProvider: [String: UsageTotals] = [:]
    for bucket in report.buckets {
        total.fold(bucket)
        byProvider[bucket.driver, default: UsageTotals()].fold(bucket)
    }
    let share = { (slice: UsageTotals) -> Double in
        total.processed > 0 ? Double(slice.processed) / Double(total.processed) : 0
    }
    // The enumerated drivers first, in their own order, then anything the wire
    // named that this build does not know — so a newer provider shows up rather
    // than vanishing into a total nothing accounts for.
    let known = usageDrivers.filter { byProvider[$0] != nil }
    let rest = byProvider.keys.filter { !usageDrivers.contains($0) }.sorted()
    return UsageFold(
        total: total,
        providers: (known + rest).compactMap { driver in
            guard let slice = byProvider[driver], slice.processed > 0 || slice.turns > 0 else { return nil }
            return UsageProviderSlice(driver: driver, totals: slice, share: share(slice))
        },
        sessions: report.sessions
    )
}

// MARK: - formatting

func formatUsd(_ value: Double) -> String { String(format: "$%.2f", value) }

/// Compact token figure — `12.3K`, `4.56M`, `1.00M`. Three significant figures,
/// the web's `formatTokens` unit for unit and rounding for rounding, so the two
/// screens print the same number the same way.
///
/// `toPrecision(3)` SPELLED OUT rather than `%.3g`, which drops the trailing
/// zeroes JavaScript keeps: a million tokens is `1.00M` on the Mac and would
/// have been `1M` here. The scaled value is always in [1, 1000) — the next unit
/// up claims anything larger — so the decimal count is just 3 minus its digits.
///
/// ONE SEAM, AT THE TOP OF EACH UNIT: a figure that rounds UP to 1000 in its own
/// unit (999,999 tokens) is `1000K` here and `1.00e+3K` on the web, because
/// `toPrecision` switches to exponential once the exponent reaches the
/// precision. Four digits is the better of the two answers and the seam is one
/// value wide per unit, so it is left alone rather than reproduced.
func formatTokens(_ value: Int) -> String {
    if value < 1_000 { return String(value) }
    for (scale, suffix) in [(1e12, "T"), (1e9, "B"), (1e6, "M"), (1e3, "K")] where Double(value) >= scale {
        let scaled = Double(value) / scale
        let decimals = scaled < 10 ? 2 : scaled < 100 ? 1 : 0
        return "\(String(format: "%.\(decimals)f", scaled))\(suffix)"
    }
    return String(value)
}

func formatShare(_ value: Double) -> String { String(format: "%.1f%%", value * 100) }

File: apps/web/lib/usage-report.ts
import type { ProviderDriverKind, TokenUsage, UsageBucket, UsageReport } from "@telar/engine-client";

/**
 * The usage page's fold — everything derivable from a `UsageReport`, derived
 * once here so the page renders figures and never arithmetic. Pure and
 * exported for tests.
 *
 * COST: the provider's own figure when it reports one, else the engine
 * prices tokens from the LiteLLM rate table (usage-pricing.ts) — cache reads
 * at the API's 0.1× rate, cache writes at 1.25× (2× for 1h TTL). A model
 * with no known rate renders as absent (`priced: false`) rather than $0.00
 * — a zero would claim the work was free.
 */

export type UsageTotals = {
  tokens: TokenUsage;
  processed: number;
  costUsd: number;
  /** True when every counted turn carried a provider cost figure. */
  priced: boolean;
  turns: number;
};

export type ProviderSlice = UsageTotals & { driver: ProviderDriverKind; share: number };
export type ModelSlice = UsageTotals & { driver: ProviderDriverKind; model: string; share: number };
export type PeriodSlice = {
  period: string;
  byDriver: Partial<Record<ProviderDriverKind, UsageTotals>>;
  total: UsageTotals;
};

export type UsageFold = {
  total: UsageTotals;
  providers: ProviderSlice[];
  models: ModelSlice[];
  /** Every period in the window, EMPTY ONES INCLUDED — a chart that skips
   *  quiet days draws a lie about the busy ones. Ascending. */
  periods: PeriodSlice[];
  sessions: number;
};

/**
 * Every provider a usage row can name.
 *
 * `telar` WAS A FOURTH until #531 removed the driver. The Agent still spends
 * tokens, and this report does not yet count them — it folds per-SESSION usage,
 * and the Agent has no session. Counting it is a separate question from
 * removing the driver, and answering it here would have meant inventing a row
 * shape nothing writes.
 */
export const DRIVERS: ProviderDriverKind[] = ["claude", "codex", "opencode"];
export const DRIVER_LABEL: Record<ProviderDriverKind, string> = { claude: "Claude", codex: "Codex", opencode: "OpenCode" };

const zeroTokens = (): TokenUsage => ({ input: 0, output: 0, cacheRead: 0, cacheCreate: 0 });
const zeroTotals = (): UsageTotals => ({ tokens: zeroTokens(), processed: 0, costUsd: 0, priced: true, turns: 0 });

export function processedTokens(tokens: TokenUsage): number {
  return tokens.input + tokens.output + tokens.cacheRead + tokens.cacheCreate;
}

function fold(into: UsageTotals, bucket: UsageBucket): void {
  into.tokens.input += bucket.tokens.input;
  into.tokens.output += bucket.tokens.output;
  into.tokens.cacheRead += bucket.tokens.cacheRead;
  into.tokens.cacheCreate += bucket.tokens.cacheCreate;
  into.processed += processedTokens(bucket.tokens);
  into.costUsd += bucket.costUsd;
  into.priced = into.priced && bucket.priced;
  into.turns += bucket.turns;
}

/** Every period the window contains, in order — days as `YYYY-MM-DD` in the
 *  report's zone, hours as epoch-ms strings, matching the buckets' own keys. */
export function windowPeriods(report: UsageReport): string[] {
  const out: string[] = [];
  if (report.resolution === "hour") {
    const HOUR = 3_600_000;
    for (let at = Math.floor(report.sinceMs / HOUR) * HOUR; at < report.untilMs; at += HOUR) out.push(String(at));
    return out;
  }
  const format = (at: number): string => {
    try {
      return new Intl.DateTimeFormat("en-CA", { timeZone: report.timeZone, year: "numeric", month: "2-digit", day: "2-digit" }).format(at);
    } catch {
      return new Intl.DateTimeFormat("en-CA", { timeZone: "UTC", year: "numeric", month: "2-digit", day: "2-digit" }).format(at);
    }
  };
  // Stepped in 24h hops from mid-window-day so a DST hour cannot skip or
  // repeat a date; dedupe guards the repeat case anyway.
  let previous = "";
  for (let at = report.sinceMs; at < report.untilMs + 86_400_000; at += 86_400_000) {
    const day = format(Math.min(at, report.untilMs));
    if (day !== previous) out.push(day);
    previous = day;
    if (at >= report.untilMs) break;
  }
  return out;
}

export function foldUsage(report: UsageReport): UsageFold {
  const total = zeroTotals();
  const byProvider = new Map<ProviderDriverKind, UsageTotals>();
  const byModel = new Map<string, ModelSlice>();
  const byPeriod = new Map<string, PeriodSlice>();
  for (const period of windowPeriods(report)) byPeriod.set(period, { period, byDriver: {}, total: zeroTotals() });

  for (const bucket of report.buckets) {
    fold(total, bucket);

    const provider = byProvider.get(bucket.driver) ?? zeroTotals();
    fold(provider, bucket);
    byProvider.set(bucket.driver, provider);

    const modelKey = `${bucket.driver}\0${bucket.model}`;
    const model = byModel.get(modelKey) ?? { ...zeroTotals(), driver: bucket.driver, model: bucket.model, share: 0 };
    fold(model, bucket);
    byModel.set(modelKey, model);

    // A bucket outside the enumerated window (clock skew, zone drift) still
    // counts in totals; it just has no column to land in.
    const period = byPeriod.get(bucket.period);
    if (period) {
      const driver = period.byDriver[bucket.driver] ?? zeroTotals();
      fold(driver, bucket);
      period.byDriver[bucket.driver] = driver;
      fold(period.total, bucket);
    }
  }

  // Shares are of PROCESSED TOKENS, not cost — cost is absent for a whole
  // provider (Codex), and a share of a figure missing half its terms would
  // always read 100% Claude.
  const share = (slice: UsageTotals): number => (total.processed > 0 ? slice.processed / total.processed : 0);

  return {
    total,
    providers: DRIVERS.flatMap((driver) => {
      const slice = byProvider.get(driver);
      return slice && (slice.processed > 0 || slice.turns > 0) ? [{ ...slice, driver, share: share(slice) }] : [];
    }),
    models: [...byModel.values()]
      .map((slice) => ({ ...slice, share: share(slice) }))
      .sort((left, right) => right.processed - left.processed),
    periods: [...byPeriod.values()],
    sessions: report.sessions,
  };
}

// ── Formatting ──────────────────────────────────────────────────────────────

export function formatUsd(value: number): string {
  return `$${value.toFixed(2)}`;
}

/** Compact token figure: `12.3K`, `4.56M` — three significant figures. */
export function formatTokens(value: number): string {
  if (value < 1_000) return String(value);
  const units: [number, string][] = [
    [1e12, "T"],
    [1e9, "B"],
    [1e6, "M"],
    [1e3, "K"],
  ];
  for (const [scale, suffix] of units) {
    if (value >= scale) {
      const scaled = value / scale;
      return `${scaled.toPrecision(3)}${suffix}`;
    }
  }
  return String(value);
}

export function formatShare(value: number): string {
  return `${(value * 100).toFixed(1)}%`;
}

/** "Aug 7" from a day key, "3 PM" from an hour key. */
export function formatPeriodShort(period: string, resolution: "day" | "hour"): string {
  if (resolution === "hour") {
    const at = Number(period);
    return Number.isFinite(at) ? new Intl.DateTimeFormat(undefined, { hour: "numeric" }).format(at) : period;
  }
  const [year, month, day] = period.split("-").map(Number);
  if (!year || !month || !day) return period;
  return new Intl.DateTimeFormat(undefined, { month: "short", day: "numeric" }).format(new Date(year, month - 1, day));
}
